src/api/general-services/work-submit/create-worker-k8s.js

```diff
@@ -63,27 +64,54 @@ const createWorkerResources = async (service) => {
 
   // Attempt to deploy the worker.
   try {
-    const params = `upgrade worker-${workerHash} chart-instance/ --namespace ${cfg.namespace} -f ${name} --install --wait -o json`.split(' ');
+    const params = `upgrade worker-${workerHash} chart-instance/ --namespace ${cfg.namespace} -f ${name} --install --atomic -o json`.split(' ');
 
     let { stdout: release } = await execFile(HELM_BINARY, params);
     release = JSON.parse(release);
 
     logger.log(`Worker instance ${release.name} successfully created.`);
   } catch (error) {
+    const params = `history worker-${workerHash} --namespace ${cfg.namespace}`;
+    logger.log(`helm update failed. Calling "helm ${params}"...`);
+    const history = await execFile(HELM_BINARY, params.split(' '));
+    logger.log(history.stdout);
+    logger.log(`If the chart is stuck updating, you may need to call "helm ${params.replace('history', 'rollback')}"`);
+
     if (!error.stderr) {
       throw error;
     }
-
     if (
       error.stderr.includes('release: already exists')
       || error.stderr.includes('another operation (install/upgrade/rollback) is in progress')
     ) {
-      logger.log('Worker instance is being created by another process, skipping...');
+      logger.log('Worker instance creation was already in progress, skipping...');
       return;
     }
 
     throw error;
   }
 };
 
+const lockHelmUpdateKey = 'lockHelmUpdate';
+const lockHelmUpdate = new AsyncLock();
+
+const createWorkerResources = async (service) => {
+  const justWait = lockHelmUpdate.isBusy(lockHelmUpdateKey);
+  if (justWait) {
+    logger.log('Helm update command lock: waiting');
+    await lockHelmUpdate.acquire(lockHelmUpdateKey, () => { logger.log('Helm update command lock: releasing'); });
+  } else {
+    logger.log('Helm update command lock: will acquire right away');
+    // eslint-disable-next-line no-async-promise-executor
+    await lockHelmUpdate.acquire(lockHelmUpdateKey, () => new Promise(async (resolve, reject) => {
```

Comment:
Correct me if Im wrong, but I think `AsyncLock` (Im looking here https://www.npmjs.com/package/async-lock) has a promise mode in which you don't need to create a promise manually as is being done here.

Attaching screenshot of an example of what I saw
<img width="853" alt="Screen Shot 2021-05-18 at 11 31 39" src="https://user-images.githubusercontent.com/8465903/118670034-a686de00-b7cc-11eb-8bcb-442b35629bd0.png">

Reply:
https://github.com/biomage-ltd/api/pull/111/commits/5c9ee5657a41fd02bc84244f74147770cd160e83 is a (non working) version of the code before I learned that I needed to return a promise. If a non-promise is returned, the lock is immediately released.

"Promise mode" is used as opposed to "callback mode" (parameter `cb` in the docs). But it is still you that need to return a promise.

Reply:
Oh so we put the promise in promise mode then lol